File: cpp/5_maze_generator.cpp
#include <algorithm>
#include <iostream>
#include <random>
#include <string>
#include <utility>
#include <vector>
using namespace std;

int main() {
    const int H = 21;
    const int W = 31;

    if (H % 2 == 0 || W % 2 == 0) {
        cerr << "H, W は奇数にしてください\n";
        return 1;
    }

    vector<string> maze(H, string(W, '#'));

    int dy[4] = {-2, 2, 0, 0};
    int dx[4] = {0, 0, -2, 2};

    random_device rd;
    mt19937 rng(rd());

    vector<pair<int, int>> st;
    st.push_back({1, 1});
    maze[1][1] = '.';

    while (!st.empty()) {
        auto [y, x] = st.back();

        vector<pair<int, int>> candidates;

        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny < 1 || ny >= H - 1 || nx < 1 || nx >= W - 1) {
                continue;
            }

            if (maze[ny][nx] == '.') {
                continue;
            }

            candidates.push_back({ny, nx});
        }

        if (candidates.empty()) {
            st.pop_back();
            continue;
        }

        shuffle(candidates.begin(), candidates.end(), rng);
        auto [ny, nx] = candidates[0];

        int wy = (y + ny) / 2;
        int wx = (x + nx) / 2;

        maze[wy][wx] = '.';
        maze[ny][nx] = '.';

        st.push_back({ny, nx});
    }

    vector<pair<int, int>> roads;

    for (int y = 0; y < H; y++) {
        for (int x = 0; x < W; x++) {
            if (maze[y][x] == '.') {
                roads.push_back({y, x});
            }
        }
    }

    shuffle(roads.begin(), roads.end(), rng);

    auto [sy, sx] = roads[0];
    auto [gy, gx] = roads[1];

    maze[sy][sx] = 'S';
    maze[gy][gx] = 'G';

    cout << H << ' ' << W << '\n';
    cout << sy << ' ' << sx << '\n';
    cout << gy << ' ' << gx << '\n';

    for (const string& line : maze) {
        cout << line << '\n';
    }

    return 0;
}

File: python/5_maze_generator.py
import random

# 奇数サイズを推奨
H = 21
W = 31

if H % 2 == 0 or W % 2 == 0:
    raise ValueError("H, W は奇数にしてください")

maze = [["#"] * W for _ in range(H)]

# 1マス飛ばしで進み、間の壁も壊す
dy = [-2, 2, 0, 0]
dx = [0, 0, -2, 2]

stack = [(1, 1)]
maze[1][1] = "."

while stack:
    y, x = stack[-1]

    candidates = []

    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]

        if not (1 <= ny < H - 1 and 1 <= nx < W - 1):
            continue

        if maze[ny][nx] == ".":
            continue

        candidates.append((ny, nx))

    if not candidates:
        stack.pop()
        continue

    ny, nx = random.choice(candidates)

    # 現在地と次のマスの間の壁を壊す
    wy = (y + ny) // 2
    wx = (x + nx) // 2

    maze[wy][wx] = "."
    maze[ny][nx] = "."

    stack.append((ny, nx))

# 通路からランダムにスタート・ゴールを選ぶ
roads = [
    (y, x)
    for y in range(H)
    for x in range(W)
    if maze[y][x] == "."
]

sy, sx = random.choice(roads)

gy, gx = random.choice(roads)
while (gy, gx) == (sy, sx):
    gy, gx = random.choice(roads)

maze[sy][sx] = "S"
maze[gy][gx] = "G"

print(H, W)
print(sy, sx)
print(gy, gx)

for line in maze:
    print("".join(line))
